pkg/upgrade/manager.go

```diff
@@ -0,0 +1,206 @@
+package upgrade
+
+// TODO consider whether this needs to exist after https://github.com/kubernetes/kubernetes/pull/79495
+// is merged, or we make the minimum supported version of k8s v1.14
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("controller").WithValues("metaKind", "upgrade")
+
+const (
+	crdName       = "constrainttemplates.templates.gatekeeper.sh"
+)
+
+// UpgradeManager allows us to upgrade resources on startup
+type UpgradeManager struct {
+	client  client.Client
+	cfg     *rest.Config
+	ctx     context.Context
+}
+
+// New creates a new manager for audit
+func New(ctx context.Context, cfg *rest.Config) (*UpgradeManager, error) {
+	am := &UpgradeManager{
+		cfg:     cfg,
+		ctx:     ctx,
+	}
+	return am, nil
+}
+
+
+// Start implements the Runnable interface
+func (um *UpgradeManager) Start(stop <-chan struct{}) error {
+	log.Info("Starting Upgrade Manager")
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	defer log.Info("Stopping upgrade manager workers")
+	errCh := make(chan error)
+	go func() {errCh <- um.upgrade(ctx)}()
+	select {
+	case <-stop:
+		return nil
+	case err := <-errCh:
+		if err != nil {
+			return err
+		}
+	}
+	// We must block indefinitely or manager will exit
+	<-stop
+	return nil
+}
+
+func (um *UpgradeManager) ensureCRDExists(ctx context.Context) error {
+	crd := &apiextensionsv1beta1.CustomResourceDefinition{}
+	return um.client.Get(ctx, types.NamespacedName{Name: crdName}, crd)
+}
+
+func (um *UpgradeManager) getAllKinds(groupVersion string) (*metav1.APIResourceList, error) {
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(um.cfg)
+	if err != nil {
+		return nil, err
+	}
+	return discoveryClient.ServerResourcesForGroupVersion(groupVersion)
+}
+
+func (um *UpgradeManager) upgrade(ctx context.Context) error {
+	gvs := []string{
+		"constraints.gatekeeper.sh/v1alpha1",
+		"templates.gatekeeper.sh/v1alpha1",
+	}
+	for _, gv := range gvs {
+		if err := um.upgradeGroupVersion(ctx, gv); err != nil {
```

Comment:
If it's a new deployment, `"constraints.gatekeeper.sh/v1alpha1"` CRs might not exist yet. This should not break the manager.

Reply:
good catch, fixed.